lib/Dancer2/Core/DSL.pm

```diff
@@ -232,11 +233,22 @@ sub _normalize_route {
     my $methods = shift;
     my %args;
 
-    # Options are optional, deduce their presence from arg length.
-    # @_ = ( REGEXP, OPTIONS, CODE )
-    # or
-    # @_ = ( REGEXP, CODE )
-    @args{qw/regexp options code/} = @_ == 3 ? @_ : ( $_[0], {}, $_[1] );
+    # Options are optional, try to deduce their presence from arg length.
+    if ( @_ == 4 ) {
+        # @_ = ( NAME, REGEXP, OPTIONS, CODE )
+        @args{qw<name regexp options code>} = @_;
+    } elsif ( @_ == 2 ) {
+        # @_ = ( REGEXP, CODE )
+        @args{qw<regexp options code>} = ( $_[0], {}, $_[1] );
+    } elsif ( @_ == 3 ) {
+        # @_ = ( REGEXP, OPTIONS, CODE )
+        # @_ = ( NAME, REGEXP, CODE )
+        if (ref $_[1] eq 'HASH') {
+            @args{qw<regexp options code>} = @_;
+        } else {
+            @args{qw<name regexp code>} = @_;
+        }
+    }
 
     return map $app->add_route( %args, method => $_ ), @{$methods};
```

Comment:
Would it be worth having add_route store a hashref per method from name to route for lookup speed and for detection of attempts to accidentally add the same route twice?

Reply:
By design, to make sure a route isn't added twice (and to locate it) we'd agree we need a hashref here. Thanks :+1:

Reply:
Dancer allows, by design, adding any amount of routes that conflict. There are two types of conflicts: Unintentional and intentional.

Unintentional would be `get '/foo' => sub {...}` and `get '/:word' => sub {...}`. A request to `/foo` can be served by both.

Intentional would be `get '/foo' => sub {...}` and `get '/foo' => sub {...}`. While the proposal for eliminating this is appealing, Dancer does allow it. (Changing this would introduce breakage of the Dancer behavior as part of this and I'm not big on doing it.)

For both cases, Dancer provides tools for picking what will happen. The first matching route will be called, then `pass` allows ending it and trying to match the next one.

```
get '/' => sub {
    # Main router
    session('user')
        or pass;
    }

    ... # handle logged in users
};

# ... other routes ...

get '/' => sub {
    template 'default';
};
```

This seems counter-intuitive, but it is the way Dancer2 is designed.

Having said this, there might be room for a strict mode that tries to find this and surface this, or even prevents it from happening, but I suggest we separate this from the current discussion so we can support what Dancer currently provides.

Reply:
Resolved